EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
-     * The destination of a jump is not considered to be accessed if it is beyond the code (`destination >= len(code)`)
-     * If code stops execution by walking past the end of the code, `PC = len(code)` is not considered to be accessed
- * If the current step of EVM execution is a `PUSH{n}`, all chunks `(PC // CHUNK_SIZE) <= chunk_index <= ((PC + n) // CHUNK_SIZE)` of the callee are accessed.
- * If a nonzero-read-size `CODECOPY` or `EXTCODECOPY` read bytes `x...y` inclusive, all chunks `(x // CHUNK_SIZE) <= chunk_index <= (min(y, code_size - 1) // CHUNK_SIZE)` of the accessed contract are accessed.
-     * Example 1: for a `CODECOPY` with start position 100, read size 50, `code_size = 200`, `x = 100` and `y = 149`
-     * Example 2: for a `CODECOPY` with start position 600, read size 0, no chunks are accessed
-     * Example 3: for a `CODECOPY` with start position 1500, read size 2000, `code_size = 3100`, `x = 1500` and `y = 3099`
- * `CODESIZE`, `EXTCODESIZE` and `EXTCODEHASH` do NOT access any chunks.
-    When a contract is created, access chunks `0 ... (len(code)+30)//31`
+##### Access events for delegation
 
-### Write Events
+When `contract_address` is that of a delegating EoA, process the following events:
 
-We define **write events** as follows. Note that when a write takes place, an access event also takes place (so the definition below should be a subset of the definition of access events). A write event is of the form `(address, sub_key, leaf_key)`, determining what data is being written to.
-
-#### Write events for account headers
-
-When a nonzero-balance-sending `CALL`, `CALLCODE` or `SELFDESTRUCT` with a given sender and recipient takes place, process these write events:
+ - For an `EXTCODEHASH`:
 
 ```
-(caller, 0, BASIC_DATA_LEAF_KEY)
-(callee, 0, BASIC_DATA_LEAF_KEY)
+(delegated_address, 0, CODEHASH_LEAF_KEY)
 ```
 
-if no account exists at `callee_address`, also process:
+ - For an `EXTCODESIZE`, or any other `*CALL`, `CALLCODE`:
 
 ```
-(callee, 0, CODEHASH_LEAF_KEY)
+(delegated_address, 0, BASIC_DATA_LEAF_KEY)
 ```
 
-When a contract creation is initialized, process these write events:
+#### Access events for code
 
+```python3
+def interpreter_loop(code: bytes, contract_address: bytes20):
+  pc = 0
+
+  # ...
+
+  while True:
+    if pc < len(code):
+      emit_code_access_events(pc, code, contract_address)
+      # ... rest of the interpreter loop
+
+# Executed when decoding an instruction when running the interpreter loop
+def emit_code_access_events(pc: int, code: bytes, contract_address: Address):
+  opcode = OP_STOP # pad chunk with 0s
+  # if the code walks beyond the code size, it is not considered to be accessed.
+  # XXX I don't think that this is true on mainnet, even before the atomic stuff!
+  if pc >= len(code):
```

Comment:
Is the basic account data loaded if we load the data from the corresponding `BASIC_DATA_LEAF_KEY`? Because then we know `len(code)`. If we know this there is no need to include a chunk which is past the code size (since we already know the code size). Same reasoning for JUMP/JUMPIs since that would jump to a non-JUMPDEST and should thus OOG without needing to include the code chunk. If `len(code)` is unknown then somehow a proof should be included that the chunk is empty.

Reply:
What about a PUSH opcode which for instance is a PUSH32 which sits at pc 16 and only has 8 bits filled? (code size is thus 24) How would this work? Mainnet would assume that the empty bytes are 0s and push that (so right-align the pushed value) and would then thus run out-of-bounds (STOP opcode) . See YP (pic) for right-padding these partially out-of-bounds PUSHs.
![image](https://github.com/user-attachments/assets/1ae0ecae-8346-43de-bdfc-6de1d9a27329)